File: methods/survival/templates/R/cutpoint.R
#!/usr/bin/env Rscript
# Template: median cut vs maximally selected log-rank (survival / maxstat).
# The naive p-value at the 'best' cut is not a valid type-I error.

suppressPackageStartupMessages({
  library(survival)
})

args <- commandArgs(trailingOnly = TRUE)
if (length(args) < 1) {
  stop("usage: cutpoint.R input.csv [time_col event_col marker_col min_frac n_perm]")
}
infile <- args[[1]]
time_col <- if (length(args) >= 2) args[[2]] else "time"
event_col <- if (length(args) >= 3) args[[3]] else "event"
marker_col <- if (length(args) >= 4) args[[4]] else "marker"
min_frac <- if (length(args) >= 5) as.numeric(args[[5]]) else 0.25
n_perm <- if (length(args) >= 6) as.integer(args[[6]]) else 1000L

df <- read.csv(infile, stringsAsFactors = FALSE)
surv <- Surv(df[[time_col]], df[[event_col]])
x <- df[[marker_col]]

med <- median(x, na.rm = TRUE)
high_med <- as.integer(x > med)
fit_med <- survdiff(surv ~ high_med)
p_med <- pchisq(fit_med$chisq, 1, lower.tail = FALSE)
cat(sprintf("median cut: chi2=%.2f p=%.4f n_high=%d n_low=%d\n",
            fit_med$chisq, p_med, sum(high_med), sum(1 - high_med)))

if (requireNamespace("maxstat", quietly = TRUE)) {
  ms <- maxstat::maxstat.test(
    surv ~ x, data = data.frame(surv = surv, x = x),
    smethod = "LogRank", pmethod = "Lau92", minprop = min_frac, maxprop = 1 - min_frac
  )
  cat(sprintf(
    "maxstat cut=%.4g statistic=%.2f p_corrected=%.4f\n",
    unname(ms$estimate), unname(ms$statistic), ms$p.value
  ))
} else {
  cat("maxstat not installed; falling back to a permutation of a grid search\n")
  qs <- quantile(x, probs = seq(min_frac, 1 - min_frac, length.out = 20), na.rm = TRUE)
  best <- list(stat = -Inf, cut = NA_real_)
  for (cut in unique(qs)) {
    g <- as.integer(x > cut)
    if (min(sum(g), sum(1 - g)) < 2) next
    st <- survdiff(surv ~ g)$chisq
    if (st > best$stat) best <- list(stat = st, cut = cut)
  }
  null <- replicate(n_perm, {
    xp <- sample(x)
    mx <- -Inf
    for (cut in unique(qs)) {
      g <- as.integer(xp > cut)
      if (min(sum(g), sum(1 - g)) < 2) next
      mx <- max(mx, survdiff(surv ~ g)$chisq)
    }
    mx
  })
  p_perm <- (1 + sum(null >= best$stat)) / (1 + n_perm)
  cat(sprintf("grid-search cut=%.4g chi2=%.2f permutation p=%.4f\n",
              best$cut, best$stat, p_perm))
}

cat("Do not report the naive p-value as the result.\n")

File: methods/survival/templates/python/cutpoint.py
#!/usr/bin/env python3
"""Template: median cut vs maximally selected log-rank cut (lifelines).

The naive p-value at the 'best' cut is not a valid type-I error. This script
reports it only next to a permutation p-value and a bootstrap of the selected
percentile. Prefer the continuous Cox model in cox_ph.py as the primary analysis.
"""

from __future__ import annotations

import argparse

import numpy as np
import pandas as pd
from lifelines import CoxPHFitter
from lifelines.statistics import logrank_test


def candidates(values: np.ndarray, min_frac: float) -> np.ndarray:
    lo, hi = np.quantile(values, [min_frac, 1 - min_frac])
    uniq = np.unique(values)
    return uniq[(uniq >= lo) & (uniq < hi)]


def max_logrank(time, event, score, min_frac):
    best = (-np.inf, np.nan)
    for cut in candidates(score, min_frac):
        high = score > cut
        if high.sum() < 2 or (~high).sum() < 2:
            continue
        res = logrank_test(time[high], time[~high], event[high], event[~high])
        if res.test_statistic > best[0]:
            best = (float(res.test_statistic), float(cut))
    return best


def hr_high(time, event, high) -> float:
    m = CoxPHFitter().fit(
        pd.DataFrame({"t": time, "e": event, "high": high.astype(int)}), "t", "e"
    )
    return float(m.summary.loc["high", "exp(coef)"])


def main() -> None:
    ap = argparse.ArgumentParser(description=__doc__)
    ap.add_argument("--input", required=True)
    ap.add_argument("--time-col", default="time")
    ap.add_argument("--event-col", default="event")
    ap.add_argument("--marker-col", default="marker")
    ap.add_argument("--min-frac", type=float, default=0.25)
    ap.add_argument("--n-perm", type=int, default=1000)
    ap.add_argument("--seed", type=int, default=20240816)
    args = ap.parse_args()

    df = pd.read_csv(args.input)
    time = df[args.time_col].to_numpy()
    event = df[args.event_col].to_numpy()
    score = df[args.marker_col].to_numpy()

    med = float(np.median(score))
    high_med = score > med
    lr_med = logrank_test(time[high_med], time[~high_med], event[high_med], event[~high_med])
    print(
        f"median cut: HR={hr_high(time, event, high_med):.2f} "
        f"log-rank p={lr_med.p_value:.4f} "
        f"n_high={high_med.sum()} n_low={(~high_med).sum()}"
    )

    stat, cut = max_logrank(time, event, score, args.min_frac)
    high = score > cut
    lr = logrank_test(time[high], time[~high], event[high], event[~high])
    rng = np.random.default_rng(args.seed)
    null = np.array(
        [
            max_logrank(time, event, score[rng.permutation(len(score))], args.min_frac)[0]
            for _ in range(args.n_perm)
        ]
    )
    p_perm = (1 + np.sum(null >= stat)) / (1 + args.n_perm)
    print(
        f"max-selected cut={cut:.4g} (percentile {(score < cut).mean()*100:.0f}): "
        f"HR={hr_high(time, event, high):.2f} "
        f"naive p={lr.p_value:.4f} permutation p={p_perm:.4f} "
        f"n_candidates={len(candidates(score, args.min_frac))}"
    )
    print("Do not report the naive p-value as the result.")


if __name__ == "__main__":
    main()
